perma_web/perma/tasks.py

```diff
@@ -2204,21 +2204,65 @@ def upload_link_to_internet_archive(link_guid, attempts=0, timeouts=0):
         logger.info(f"Queued Link {link_guid} no longer eligible for upload.")
         return
 
-    # Make sure we've already created the IA item for this link
+    # Get or create the appropriate InternetArchiveItem object for this link
+    date_string = link.creation_timestamp.strftime('%Y-%m-%d')
     identifier = InternetArchiveItem.DAILY_IDENTIFIER.format(
         prefix=settings.INTERNET_ARCHIVE_DAILY_IDENTIFIER_PREFIX,
-        date_string=link.creation_timestamp.strftime('%Y-%m-%d')
+        date_string=date_string
     )
-    try:
-        perma_item = InternetArchiveItem.objects.get(identifier=identifier)
-    except InternetArchiveItem.DoesNotExist:
-        msg = f"Internet Archive Item {identifier} does not exist: not uploading {link_guid}."
-        if settings.INTERNET_ARCHIVE_EXCEPTION_IF_NO_ITEM:
-            logger.error(msg)
+    start = InternetArchiveItem.datetime(f"{date_string} 00:00:00")
+    end = start + timedelta(days=1)
+    perma_item, _created = InternetArchiveItem.objects.get_or_create(
+        identifier=identifier,
+        span=(start, end),
+    )
+
+    # Check the Internet Archive-related status of this link
+    perma_file = InternetArchiveFile.objects.filter(
+        item_id=identifier,
+        link_id=link_guid
+    ).first()
+    if perma_file:
+        if perma_file.status == 'confirmed_present':
+            logger.info(f"Not uploading {link_guid} to {identifier}: our records indicate it is already present.")
+            return
+        elif perma_file.status == 'deletion_attempted':
+            # If we find ourselves here, something has gotten very mixed up indeed. We probably need a human to have a look.
+            # Use the error log, assuming this will happen rarely or never.
+            logger.error(f"Please investigate the status of {link_guid}: our records indicate a deletion attempt is in progress, but an upload was attempted in the meantime.")
```

Comment:
The pattern I learned for Celery tasks was, work hard to avoid uncaught exceptions, so that the task itself finishes and is a "success" so far as Celery is concerned. Kind of like... if an uncaught exception is thrown, it's the equivalent of an `E` in a pytest test: something has gone wrong with the code, rather than, the task/test "failed" with an `F`.

It may be that is not the best pattern.

Thanks for the nudge. Will definitely read about that, as well as [returning and storing results](https://github.com/harvard-lil/perma/pull/3259#discussion_r1105850823) as per your other comment.

Reply:
I think it just depends on the semantics of the error—the comment made me think this was a real bad unexpected state that I'd want to be alerted about.

Reply:
I'm a little less sure how this works now that we are using Sentry. In the olden times, writing to `logger.error` like this would send us error emails, just like raising an exception would.